### PR TITLE
DataGrid - Add the editing.newRowPosition option

### PR DESCRIPTION
**Pull request by Alyar666**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 1]
```diff
@@ -33,6 +33,11 @@ import {
     EDIT_MODE_ROW,
     EDIT_MODES,
     ROW_BASED_MODES,
+    FIRST_NEW_ROW_POSITION,
+    LAST_NEW_ROW_POSITION,
+    PAGE_BOTTOM_NEW_ROW_POSITION,
+    PAGE_TOP_NEW_ROW_POSITION,
+    VIEWPORT_BOTTOM_NEW_ROW_POSITION
 } from './ui.grid_core.editing_constants';
```
Comment (pomahtri) [inline js/ui/grid_core/ui.grid_core.editing.js]:
```suggestion
    VIEWPORT_BOTTOM_NEW_ROW_POSITION,
    VIEWPORT_TOP_NEW_ROW_POSITION
} from './ui.grid_core.editing_constants';
```

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 2]
```diff
@@ -2082,11 +2159,10 @@ const EditingController = modules.ViewController.inherit((function() {
 export const editingModule = {
     defaultOptions: function() {
         return {
-
-
             editing: {
                 mode: 'row', // "batch"
                 refreshMode: 'full',
+                newRowPosition: 'viewportTop',
```
Comment (pomahtri) [inline js/ui/grid_core/ui.grid_core.editing.js]:
```suggestion
                newRowPosition: VIEWPORT_TOP_NEW_ROW_POSITION,
```

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 3]
```diff
@@ -654,21 +674,25 @@ const EditingController = modules.ViewController.inherit((function() {
             changes.forEach(change => {
                 const isInsert = change.type === DATA_EDIT_DATA_INSERT_TYPE;
 
-                if(!isInsert) {
+                if(!isInsert || this.needChangePageIndexToInsertRow()) {
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js:677]: User can combine insertAfterKey and newRowPosition: 'bottom'. And in this case insertAfterKey should not be ignored. You can fix it in another PR.

**Review by NickMitrokhin — COMMENTED** [review]
Found issues:
1. insertBeforeKey or insertAfterKey is ignored when newRowPosition is specified.
2. If the viewport is scrolled to the top and a new row is inserted using the "+" button in the 'last' newRowPosition, the viewport is not scrolled to the new row (virtual scrolling).
3. If the viewport is scrolled to the bottom and a new row is inserted using the "+" button in the 'first' newRowPosition, the viewport is not scrolled to the new row (virtual scrolling).

**Review by vconst — APPROVED** [review]

**Comment by pomahtri** [comment]
Am I right that we can't use both `insertAfterKey` and `insertBeforeKey`? If yes, we can change typing to prevent this. Don't know if we should do this.

[There is sample](https://www.typescriptlang.org/play?ts=4.3.5#code/C4TwDgpgBAIghsOBhAFnAdgc2gXigbwCgoSoBrCEALigxAG5jTRIaByAS3QGcIAnYGygAfKGwCuYACYIIQ0Wz4QAtgHsAbnMakoMxDVUAjAFYQAxsG2kuUiAA8A-DXTjlh-lZJg42AJLpbR2dXdz5GAF8oADIoAAoiEi5eAQBBADNgfgBpSidadAYmJP5gACEINNUlHJA89AhNMMJI0QSoYtSM7NznBo8inhLyyuqe-MLwgEpCQjNVHmBdBGQ0LAgkKqULGnhEVAxsKDw2imooAEYAGiYWCHYOwWuSPTgafHCn9sGBYaqIGpo52aMzmCyWe1W2AA6nx5pgdst9msjgQmKdAZ9bvdvo8mC83h8BskyhU-gCLp8HulMnxyVdmkA)

**Comment by Alyar666** [comment]
> Am I right that we can't use both `insertAfterKey` and `insertBeforeKey`? If yes, we can change typing to prevent this. Don't know if we should do this.
> 
> [There is sample](https://www.typescriptlang.org/play?ts=4.3.5#code/C4TwDgpgBAIghsOBhAFnAdgc2gXigbwCgoSoBrCEALigxAG5jTRIaByAS3QGcIAnYGygAfKGwCuYACYIIQ0Wz4QAtgHsAbnMakoMxDVUAjAFYQAxsG2kuUiAA8A-DXTjlh-lZJg42AJLpbR2dXdz5GAF8oADIoAAoiEi5eAQBBADNgfgBpSidadAYmJP5gACEINNUlHJA89AhNMMJI0QSoYtSM7NznBo8inhLyyuqe-MLwgEpCQjNVHmBdBGQ0LAgkKqULGnhEVAxsKDw2imooAEYAGiYWCHYOwWuSPTgafHCn9sGBYaqIGpo52aMzmCyWe1W2AA6nx5pgdst9msjgQmKdAZ9bvdvo8mC83h8BskyhU-gCLp8HulMnxyVdmkA)

You can use both insertAfterKey and insertBeforeKey. In this case, insertBeforeKey will be ignored.

**Pull request merged**